controller/test/frame_detector/frame_detector_test.cpp

```diff
@@ -0,0 +1,236 @@
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  for (uint32_t i = 0; i < LEN; i++) {
+    rx_buf.PutByte(' ');
+  }
+
+  EXPECT_EQ(State::LOST, frame_detector.get_state());
+  ASSERT_FALSE(frame_detector.is_frame_available());
+}
+
+TEST(FrameDetector, RxFrameIllegalyLong) {
+  FrameDetectorTest frame_detector(rx_buf);
+  EXPECT_TRUE(frame_detector.Begin());
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  for (uint32_t i = 0; i < LEN - 1; i++) {
+    rx_buf.PutByte(' ');
+  }
+  rx_buf.PutByte(MARK);
+
+  EXPECT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  ASSERT_FALSE(frame_detector.is_frame_available());
+}
+
+TEST(FrameDetector, ErrorWhileRx) {
+  FrameDetectorTest frame_detector(rx_buf);
+  EXPECT_TRUE(frame_detector.Begin());
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  frame_detector.OnRxError(RxError::TIMEOUT);
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  frame_detector.OnRxError(RxError::DMA);
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  frame_detector.OnRxError(RxError::OVERFLOW);
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  frame_detector.OnRxError(RxError::SERIAL_FRAMING);
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  frame_detector.OnRxError(RxError::UNKNOWN);
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+}
+
+vector<string> fakeRx(const char *frame, FrameDetectorTest &frame_detector) {
+  vector<string> ret;
+  for (uint32_t i = 0; i < strlen(frame); i++) {
+    rx_buf.PutByte(frame[i]);
+    if (frame_detector.is_frame_available()) {
+      string s;
+      s.assign(reinterpret_cast<char *>(frame_detector.TakeFrame()),
```

Comment:
This should be split in two statements - TakeFrame() is a destructive operation and one can imagine get_frame_length() failing if you call it when the frame is already unavailable.

Reply:
done